hibernate-core/src/test/java/org/hibernate/jpa/test/hibernateFilters/ProxyPreservingFiltersOutsideInitialSessionTest.java

```diff
@@ -71,4 +72,179 @@ public void testPreserveFilters() {
+					return accountGroup;
+				} );
+	}
+
+	@Test
+	public void testChangeFilterBeforeInitializeInTempSession() {
+
+		doInJPA(
+				this::entityManagerFactory, entityManager -> {
+					AccountGroup accountGroup = new AccountGroup();
+					accountGroup.setId( 1L );
+					entityManager.persist( accountGroup );
+
+					Account account = new Account();
+					account.setName( "A1" );
+					account.setRegionCode( "Europe" );
+					entityManager.persist( account );
+					accountGroup.getAccounts().add( account );
+
+					account = new Account();
+					account.setName( "A2" );
+					account.setRegionCode( "Europe" );
+					entityManager.persist( account );
+					accountGroup.getAccounts().add( account );
+
+					account = new Account();
+					account.setName( "A3" );
+					account.setRegionCode( "US" );
+					entityManager.persist( account );
+					accountGroup.getAccounts().add( account );
+				}
+		);
+
+		AccountGroup group = doInJPA( this::entityManagerFactory, entityManager -> {
+					entityManager.unwrap( Session.class ).enableFilter( "byRegion" ).setParameter( "region", "US" );
+					AccountGroup accountGroup = entityManager.find( AccountGroup.class, 1L );
+					// Change the filter.
+					entityManager.unwrap( Session.class ).enableFilter( "byRegion" ).setParameter( "region", "Europe" );
+					return accountGroup;
+				} );
+
+		// What should group.getAccounts() contain? Should it be accounts with regionCode "Europe"
+		// because that was the most recent filter used in the session?
+		Hibernate.initialize( group.getAccounts() );
+		// The following will fail because the collection will only contain accounts with regionCode "US"
+		assertEquals(2, group.getAccounts().size());
+	}
+
+	@Test
+	public void testMergeNoFilterThenInitializeTempSession() {
+
+		doInJPA(
+				this::entityManagerFactory, entityManager -> {
+					AccountGroup accountGroup = new AccountGroup();
+					accountGroup.setId( 1L );
+					entityManager.persist( accountGroup );
+
+					Account account = new Account();
+					account.setName( "A1" );
+					account.setRegionCode( "Europe" );
+					entityManager.persist( account );
+					accountGroup.getAccounts().add( account );
+
+					account = new Account();
+					account.setName( "A2" );
+					account.setRegionCode( "Europe" );
+					entityManager.persist( account );
+					accountGroup.getAccounts().add( account );
+
+					account = new Account();
+					account.setName( "A3" );
+					account.setRegionCode( "US" );
+					entityManager.persist( account );
+					accountGroup.getAccounts().add( account );
+				}
+		);
+
+		final AccountGroup group = doInJPA( this::entityManagerFactory, entityManager -> {
+					entityManager.unwrap( Session.class ).enableFilter( "byRegion" ).setParameter( "region", "US" );
+					return entityManager.find( AccountGroup.class, 1L );
+				} );
+
+		final AccountGroup mergedGroup = doInJPA( this::entityManagerFactory, entityManager -> {
+					return entityManager.merge( group );
+				} );
+
```

Comment:
Compare the following behavior to what happens when the collection is initialized after Session#saveOrUpdate in testSaveOrUpdateNoFilterThenInitializeTempSession().